### PR TITLE
Add udp and dns support 添加udp和dns支持 

### PR DESCRIPTION
**Pull request by reflyable**
接入smoltcp udp和dns socket, 实现对应udp栈和dns相关的Trait，增加相应rust app

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,12 +7,14 @@ members = [
     "apps/memtest",
     "apps/fs/shell",
     "apps/net/echoserver",
-    "apps/net/httpclient",
+    # "apps/net/httpclient",
```
Comment (equation314) [inline Cargo.toml]: Uncomment it, and remove line 13

File: Cargo.toml [hunk 2]
```diff
@@ -7,12 +7,14 @@ members = [
...
+    "apps/net/udpserver",
     "apps/net/httpserver",
+    "apps/net/httpclient",
     "apps/task/parallel",
     "apps/task/sleep",
     "apps/task/yield",
-
+
```
Comment (equation314) [inline Cargo.toml]: Remove the Tab

File: apps/net/httpclient/Cargo.toml [hunk 3]
```diff
@@ -2,9 +2,12 @@
...
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-libax = { path = "../../../ulib/libax", features = ["paging", "net"] }
+libax = { path = "../../../ulib/libax", features = ["paging", "multitask", "net"] }
+
+[features]
+default = []
+dns = []
```
Comment (equation314) [inline apps/net/httpclient/Cargo.toml]: Add blank line at the end of file

File: apps/net/httpclient/src/main.rs [hunk 4]
```diff
@@ -4,27 +4,71 @@
...
-use libax::net::{IpAddr, TcpStream};
+use libax::net::{SocketAddr, TcpStream, ToSocketAddrs};
 
-const DEST_IP: &str = "49.12.234.183"; // ident.me
-const REQUEST: &str = "\
-GET / HTTP/1.1\r\n\
-Host: ident.me\r\n\
-Accept: */*\r\n\
-\r\n";
+const DEST_HOST: &str = "www.example.com";
```
Comment (equation314) [inline apps/net/httpclient/src/main.rs]: Why use "www.example.com" instead of "ident.me"?

File: apps/net/httpclient/Cargo.toml [hunk 5]
```diff
@@ -2,9 +2,12 @@
 name = "arceos-httpclient"
 version = "0.1.0"
 edition = "2021"
-authors = ["Yuekai Jia <equation618@gmail.com>"]
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-libax = { path = "../../../ulib/libax", features = ["paging", "net"] }
+libax = { path = "../../../ulib/libax", features = ["paging", "multitask", "net"] }
```
Comment (equation314) [inline apps/net/httpclient/Cargo.toml]: Why use  "multitask"?

File: apps/net/httpclient/Cargo.toml [hunk 6]
```diff
@@ -2,9 +2,12 @@
 name = "arceos-httpclient"
 version = "0.1.0"
 edition = "2021"
-authors = ["Yuekai Jia <equation618@gmail.com>"]
```
Comment (equation314) [inline apps/net/httpclient/Cargo.toml:5]: Add your email here

File: apps/net/udpserver/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                println!("{}", mid);
+                let mid = ["response_", mid].join("");
+                socket.send_to(mid.as_bytes(), addr)?;
+                buf = [0u8; 1024];
+            }
+            Err(e) => return Err(e),
+        };
+    }
+}
+#[no_mangle]
```
Comment (equation314) [inline apps/net/udpserver/src/main.rs:36]: Add blank line

File: modules/axnet/Cargo.toml [hunk 8]
```diff
@@ -31,5 +31,5 @@ features = [
   "alloc", "log",   # no std
   "medium-ethernet",
   "proto-ipv4",
-  "socket-raw", "socket-icmp", "socket-udp", "socket-tcp",
+  "socket-raw", "socket-icmp", "socket-udp", "socket-tcp","socket-dns"
```
Comment (equation314) [inline modules/axnet/Cargo.toml]: Add a space before "socket-dns"

File: apps/net/udpserver/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+use libax::io;
+use libax::net::{IpAddr, UdpSocket};
+
+const LOCAL_IP: &str = "10.0.2.15";
+const LOCAL_PORT: u16 = 5555;
+
+fn accept_loop() -> io::Result {
+    let (addr, port) = (IpAddr::from_str(LOCAL_IP).unwrap(), LOCAL_PORT);
+    let socket = UdpSocket::bind((addr, port).into())?;
+    println!("listen on: http://{}/", socket.local_addr().unwrap());
```
Comment (equation314) [inline apps/net/udpserver/src/main.rs]: Do not use `http://`

File: apps/net/udpserver/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+use core::str::FromStr;
+
+use libax::io;
+use libax::net::{IpAddr, UdpSocket};
+
+const LOCAL_IP: &str = "10.0.2.15";
+const LOCAL_PORT: u16 = 5555;
+
+fn accept_loop() -> io::Result {
```
Comment (equation314) [inline apps/net/udpserver/src/main.rs]: UDP socket does not use `accept()`, rename the function, e.g., `receive_loop()`

File: apps/net/httpclient/src/main.rs [hunk 11]
```diff
@@ -4,27 +4,71 @@
...
-    let response = core::str::from_utf8(&buf[..n]).unwrap();
-    println!("{}", response);
+    let requset: String =
+        "GET / HTTP/1.1\r\nHost: ".to_string() + DEST_HOST + "\r\nAccept: */*\r\n\r\n";
+
+    let mut stream = TcpStream::connect(get_addr())?;
+    stream.write(requset.as_bytes())?;
+    let mut buf = [0; 2048];
+    let mut length;
+    {
```
Comment (equation314) [inline apps/net/httpclient/src/main.rs]: I think the original version is fine, this code leads to the example is not simple enough.

File: apps/net/udpserver/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "arceos-udpserver"
+version = "0.1.0"
+edition = "2021"
+authors = ["Dashuai Wu"]
```
Comment (equation314) [inline apps/net/udpserver/Cargo.toml]: Add your email

File: ulib/libax/src/net/udp.rs [hunk 13]
```diff
@@ -0,0 +1,28 @@
+use crate::io;
+use axnet::SocketAddr;
+use axnet::UdpSocket as _UdpSocket;
+/// A TCP socket.
```
Comment (equation314) [inline ulib/libax/src/net/udp.rs]: Fix this, and add a blank line before the comment and `use ...`

File: ulib/libax/src/net/udp.rs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+impl UdpSocket {
+    /// Creates a UDP socket from the given address.
+    pub fn bind(addr: SocketAddr) -> io::Result<Self> {
+        let mut socket = _UdpSocket::new();
+        socket.bind(addr)?;
+        Ok(Self { socket })
+    }
+    /// Returns the socket address that this socket was created from.
```
Comment (equation314) [inline ulib/libax/src/net/udp.rs:18]: Add blank lines between two functions.

File: ulib/libax/src/net/udp.rs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.socket.local_addr()
+    }
+    ///Receives a message on the socket.
```
Comment (equation314) [inline ulib/libax/src/net/udp.rs]: Add a space after `///`

File: ulib/libax/src/net/udp.rs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    pub fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
+        self.socket.recvfrom(buf)
+    }
+    ///Sends data on the socket to the given address.
```
Comment (equation314) [inline ulib/libax/src/net/udp.rs]: Add a space after `///`

File: ulib/libax/src/net/socket_addr.rs [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+use crate::io;
+
+use axnet::DnsSocket;
+
+fn resolve_socket_addr(name: &str) -> io::Result<alloc::vec::Vec<IpAddr>> {
+    let socket = DnsSocket::new();
+    socket.query(name, axnet::DnsQueryType::A)
+}
+/// A trait for objects which can be converted or resolved to one or more SocketAddr values.
```
Comment (equation314) [inline ulib/libax/src/net/socket_addr.rs]: use ``[`SocketAddr`]``, add a blank line before the comment.

File: ulib/libax/src/net/socket_addr.rs [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+pub trait ToSocketAddrs {
+    ///Returned iterator over socket addresses which this type may correspond to.
+    type Iter: Iterator<Item = SocketAddr>;
+    ///Converts this object to an iterator of resolved SocketAddrs.
```
Comment (equation314) [inline ulib/libax/src/net/socket_addr.rs]: use [`SocketAddr`], add a blank line before the comment.

File: ulib/libax/src/net/socket_addr.rs [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    type Iter = vec::IntoIter<SocketAddr>;
+    fn to_socket_addrs(&self) -> io::Result<vec::IntoIter<SocketAddr>> {
+        let (host, port) = *self;
+
+        // try to parse the host as a regular IP address first
+        if let Ok(addr) = host.parse::<IpAddr>() {
+            let addr = SocketAddr::new(addr, port);
+            return Ok(vec![addr].into_iter());
+        }
+        // if let Ok(addr) = host.parse::<Ipv6Addr>() {
```
Comment (equation314) [inline ulib/libax/src/net/socket_addr.rs]: Remove the unused code

File: ulib/libax/src/net/udp.rs [hunk 20]
```diff
@@ -0,0 +1,28 @@
+use crate::io;
+use axnet::SocketAddr;
+use axnet::UdpSocket as _UdpSocket;
+/// A TCP socket.
+pub struct UdpSocket {
+    socket: _UdpSocket,
```
Comment (equation314) [inline ulib/libax/src/net/udp.rs]: Use `axnet::UdpSocket` here,

File: ulib/libax/src/net/socket_addr.rs [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+use axerrno;
+use axerrno::ax_err_type;
+use axnet::{IpAddr, SocketAddr};
+
+use crate::io;
+
+use axnet::DnsSocket;
+
+fn resolve_socket_addr(name: &str) -> io::Result<alloc::vec::Vec<IpAddr>> {
```
Comment (equation314) [inline ulib/libax/src/net/socket_addr.rs]: Move this function into `axnet`

File: modules/axnet/src/lib.rs [hunk 22]
```diff
@@ -31,8 +31,12 @@ cfg_if::cfg_if! {
     }
 }
 
+pub use self::net_impl::DnsSocket;
```
Comment (equation314) [inline modules/axnet/src/lib.rs]: Do not export `DnsSocket`

File: modules/axnet/src/lib.rs [hunk 23]
```diff
@@ -31,8 +31,12 @@ cfg_if::cfg_if! {
...
 pub use self::net_impl::TcpSocket;
-pub use smoltcp::wire::{IpAddress as IpAddr, IpEndpoint as SocketAddr, Ipv4Address as Ipv4Addr};
+pub use self::net_impl::UdpSocket;
+pub use smoltcp::wire::{
+    DnsQueryType, IpAddress as IpAddr, IpEndpoint as SocketAddr, Ipv4Address as Ipv4Addr,
```
Comment (equation314) [inline modules/axnet/src/lib.rs]: Do not export `DnsQueryType`

File: modules/axnet/src/smoltcp_impl/dns.rs [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+use smoltcp::iface::SocketHandle;
+use smoltcp::socket::dns::{self, GetQueryResultError, StartQueryError};
+use smoltcp::wire::{DnsQueryType, IpAddress};
+
+use super::{SocketSetWrapper, ETH0, SOCKET_SET};
+
+/// A DNS socket to simplify DNS query.
+///
+
+/// [`update_servers`]: UDPSocket::update_servers
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/dns.rs]: Remove the tow comments

File: modules/axnet/src/smoltcp_impl/mod.rs [hunk 25]
```diff
@@ -18,17 +20,21 @@ use smoltcp::wire::{EthernetAddress, HardwareAddress, IpAddress, IpCidr};
 
 use self::listen_table::ListenTable;
 
+pub use self::dns::DnsSocket;
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/mod.rs]: Do not export `DnsSocket`

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+
+use smoltcp::iface::SocketHandle;
+use smoltcp::socket::udp::{self, BindError, SendError};
+
+use super::{SocketSetWrapper, SOCKET_SET};
+use crate::SocketAddr;
+/// A UDP socket that provides POSIX-like APIs.
+///
+
+/// [`bind`]: UDPSocket::bind
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs]: Remove these comments

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+    /// [`Err(NotConnected)`](AxError::NotConnected) if not connected.
+    pub fn local_addr(&self) -> AxResult<SocketAddr> {
+        self.local_addr.ok_or(AxError::NotConnected)
+    }
+    /// Binds an unbound socket to the given address and port.
+    ///
+    /// It's must be called before [`sendto`](Self::sendto) and
+    /// [`recvfrom`](Self::recvfrom).
+    pub fn bind(&mut self, addr: SocketAddr) -> AxResult {
+        debug!("{:?}", addr);
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs]: Add more detail debug info ,or just remove it

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+        let handle = self
+            .handle
+            .ok_or_else(|| ax_err_type!(InvalidInput, "socket bind() failed"))?;
+        if self.local_addr.is_some() {
+            return ax_err!(InvalidInput, "socket bind() failed: already bound");
+        }
+        debug!("{:?}", addr);
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs]: Same as above

File: modules/axnet/src/smoltcp_impl/mod.rs [hunk 29]
```diff
@@ -254,6 +275,7 @@ fn snoop_tcp_packet(buf: &[u8]) -> Result<(), smoltcp::wire::Error> {
             // create a socket for the first incoming TCP packet, as the later accept() returns.
             LISTEN_TABLE.incoming_tcp_packet(src_addr, dst_addr);
         }
+    } else if ipv4_packet.next_header() == IpProtocol::Udp {
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/mod.rs]: Remove this `else if`

**Review by equation314 — CHANGES_REQUESTED** [review]

File: apps/net/httpclient/src/main.rs (hunk 4 above)
Comment (reflyable) [reply apps/net/httpclient/src/main.rs]: ident.me的reply太短，原有的实现方式在获得的包长大于1500(MTU)时会截断，在原来的httpclient用www.example.com可以复现这个问题。"www.example.com"也能更好地体现是一个http response

File: modules/axnet/src/smoltcp_impl/dns.rs [hunk 30]
```diff
@@ -0,0 +1,85 @@
+use alloc::vec::Vec;
+use axerrno::{ax_err_type, AxError, AxResult};
+
+use smoltcp::iface::SocketHandle;
+use smoltcp::socket::dns::{self, GetQueryResultError, StartQueryError};
+use smoltcp::wire::{DnsQueryType, IpAddress};
+
+use super::{SocketSetWrapper, ETH0, SOCKET_SET};
+
+/// [`query`]: UDPSocket::query
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/dns.rs]: Fix the comment

File: modules/axnet/src/smoltcp_impl/mod.rs [hunk 31]
```diff
@@ -60,6 +66,21 @@ impl<'a> SocketSetWrapper<'a> {
         let tcp_tx_buffer = socket::tcp::SocketBuffer::new(vec![0; TCP_TX_BUF_LEN]);
         socket::tcp::Socket::new(tcp_rx_buffer, tcp_tx_buffer)
     }
+    pub fn new_udp_socket() -> socket::udp::Socket<'a> {
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/mod.rs:70]: Add a new line before

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+use smoltcp::iface::SocketHandle;
+use smoltcp::socket::udp::{self, BindError, SendError};
+
+use super::{SocketSetWrapper, SOCKET_SET};
+use crate::SocketAddr;
+
+/// [`sendto`]: UDPSocket::sendto
+/// [`recvfrom`]: UDPSocket::recvfrom
+/// [`peekfrom`]: UDPSocket::peekfrom
+pub struct UdpSocket {
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs:10]: Fix the unclear comments

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 33]
```diff
@@ -0,0 +1,181 @@
...
+                BindError::Unaddressable => {
+                    ax_err!(InvalidInput, "socket bind() failed")
+                }
+            })?;
+            Ok(socket.endpoint())
+        })?;
+        self.local_addr = Some(addr);
+        Ok(())
+    }
+    /// Transmits data in the given buffer to the given address.
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs:59]: Add a new line before

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 34]
```diff
@@ -0,0 +1,181 @@
...
+                    SOCKET_SET.poll_interfaces();
+                    return Ok(n);
+                }
+                Err(AxError::Again) => axtask::yield_now(),
+                Err(e) => return Err(e),
+            }
+        }
+    }
+    /// Receives data from the socket, stores it in the given buffer.
+    pub fn recvfrom(&self, buf: &mut [u8]) -> AxResult<(usize, SocketAddr)> {
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs:95]: Add a new line before

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 35]
```diff
@@ -0,0 +1,181 @@
...
+                    SOCKET_SET.poll_interfaces();
+                    return Ok(x);
+                }
+                Err(AxError::Again) => axtask::yield_now(),
+                Err(e) => return Err(e),
+            }
+        }
+    }
+    /// Close the socket.
+    pub fn shutdown(&self) -> AxResult {
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs:128]: Add a new line before

File: modules/axnet/src/smoltcp_impl/udp.rs [hunk 36]
```diff
@@ -0,0 +1,181 @@
...
+            SOCKET_SET.with_socket_mut::<udp::Socket, _, _>(handle, |socket| {
+                debug!("socket {}: shutting down", handle);
+                socket.close();
+            });
+        } else {
+            return ax_err!(InvalidInput, "socket shutdown() failed");
+        }
+        Ok(())
+    }
+    /// Receives data from the socket, stores it in the given buffer, without removing it from the queue.
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/udp.rs:142]: Add a new line before

File: ulib/libax/src/net/socket_addr.rs [hunk 37]
```diff
@@ -0,0 +1,96 @@
...
+
+use crate::io;
+
+use axnet::resolve_socket_addr;
+
+/// A trait for objects which can be converted or resolved to one or more [`SocketAddr`] values.
+pub trait ToSocketAddrs {
+    /// Returned iterator over socket addresses which this type may correspond to.
+    type Iter: Iterator<Item = SocketAddr>;
+    /// Converts this object to an iterator of resolved [`SocketAddr`]s.
```
Comment (equation314) [inline ulib/libax/src/net/socket_addr.rs:22]: Add a new line before

File: apps/net/httpclient/src/main.rs [hunk 38]
```diff
@@ -4,28 +4,37 @@
...
+}
 
 fn client() -> io::Result {
-    let (addr, port) = (IpAddr::from_str(DEST_IP).unwrap(), 80);
-    let mut stream = TcpStream::connect((addr, port).into())?;
-    stream.write(REQUEST.as_bytes())?;
-
-    let mut buf = [0; 1024];
+    let requset: String =
+        "GET / HTTP/1.1\r\nHost: ".to_string() + DEST_HOST + "\r\nAccept: */*\r\n\r\n";
```
Comment (equation314) [inline apps/net/httpclient/src/main.rs]: Use the original `REQUEST` constant.

File: apps/net/httpclient/expect_info.out [hunk 39]
```diff
@@ -25,13 +25,16 @@ created net interface "eth0":
...
 Primary CPU 0 init OK.
 Hello, simple http client!
+IP:
+
 HTTP/1.1 200 OK
 Server: nginx
 Date:
 Content-Type: text/plain
-Content-Length:
+Content-Length: 14
```
Comment (equation314) [inline apps/net/httpclient/expect_info.out]: Remove `14`, or other IP may cause test failure.

File: apps/net/httpclient/expect_info_dns.out [hunk 40]
```diff
@@ -0,0 +1,39 @@
...
+  gateway:  10.0.2.2
+Initialize interrupt handlers...
+Primary CPU 0 init OK.
+Hello, simple http client!
+ident.me IP:
+HTTP/1.1 200 OK
+Server: nginx
+Date:
+Content-Type: text/plain
+Content-Length: 14
```
Comment (equation314) [inline apps/net/httpclient/expect_info_dns.out]: Remove `14`

File: scripts/test/app_test.sh
Comment (equation314) [inline scripts/test/app_test.sh:1]: Do not change this file

**Review by equation314 — CHANGES_REQUESTED** [review]

File: apps/net/udpserver/src/main.rs [hunk 41]
```diff
@@ -0,0 +1,40 @@
...
+
+use core::str::FromStr;
+
+use libax::io;
+use libax::net::{IpAddr, UdpSocket};
+
+const LOCAL_IP: &str = "10.0.2.15";
+const LOCAL_PORT: u16 = 5555;
+
+fn recive_loop() -> io::Result {
```
Comment (equation314) [inline apps/net/udpserver/src/main.rs]: typo

File: apps/net/udpserver/src/main.rs [hunk 42]
```diff
@@ -0,0 +1,40 @@
...
+    let (addr, port) = (IpAddr::from_str(LOCAL_IP).unwrap(), LOCAL_PORT);
+    let socket = UdpSocket::bind((addr, port).into())?;
+    println!("listen on: {}/", socket.local_addr().unwrap());
```
Comment (equation314) [inline apps/net/udpserver/src/main.rs]: Remove the `/`

File: modules/axnet/src/smoltcp_impl/dns.rs [hunk 43]
```diff
@@ -0,0 +1,86 @@
+use alloc::vec::Vec;
+use axerrno::{ax_err_type, AxError, AxResult};
+
+use smoltcp::iface::SocketHandle;
+use smoltcp::socket::dns::{self, GetQueryResultError, StartQueryError};
+use smoltcp::wire::{DnsQueryType, IpAddress};
+
+use super::{SocketSetWrapper, ETH0, SOCKET_SET};
+
+/// A Dns socket.
```
Comment (equation314) [inline modules/axnet/src/smoltcp_impl/dns.rs]: Use uppercase: `DNS`

File: modules/axnet/src/lib.rs
Comment (equation314) [inline modules/axnet/src/lib.rs:1]: Update crate-level documents: line 3, line 9-11

**Review by equation314 — CHANGES_REQUESTED** [review]

**Review by equation314 — APPROVED** [review]

**Pull request merged**